### ISSUE TITLE
`Badarg` error when using `file2tab`

### ISSUE THREAD

**Issue by ofekd** [Michael-Mark-Edu/bravo#2]
Hello

Here's the output:

```
runtime error: Erlang error

An error occurred outside of Gleam.

erlang:error(Badarg)

stacktrace:
  ets.first unknown source
  bravo_ffi.try_file2tab /<redacted>/build/dev/erlang/bravo/_gleam_artefacts/bravo_ffi.erl:95
  bravo/uset.file2tab /<redacted>/build/packages/bravo/src/bravo/uset.gleam:158
  sample.main /<redacted>/sample.gleam:53
```

The code that throws:

``` gleam
  let path = server_priv |> filepath.join("hello.dat")
  let _ =
    uset.new("hello", 1, bravo.Public)
    |> result.then(fn(table) {
      table |> uset.insert([#("Hello", "world")])
      uset.tab2file(table, path, False, False, False)
    })

  let _ =
    uset.file2tab(path, False, dynamic.tuple2(dynamic.string, dynamic.string))
```

**Comment by ofekd** [comment]
This is working fine in `master`, is there any reason it's not deployed? Wondering if there are problems to be aware of

**Comment by Michael-Mark-Edu** [comment]
I got burned out on this project a few months ago and haven't touched it since. 

The master branch contains the partially- complete v5.0.0 update and I might have fixed this issue there? I was planning on releasing everything as a big update, but ultimately that fell through...

I'm really busy with college for the next few days, so I probably won't have time to figure this issue out, but I'll keep it in mind. If I do come back, I might post a v4.0.2 that fixes this issue specifically and continue working on v5.0.0 if motivation returns to me, but that's all a big if.

**Comment by ofekd** [comment]
Very understandable

I've took version 5 for a spin -- I've vendored it it's been smooth so far. It's a very pleasant API, well done.

Maybe before you do actual work, can you make a 5.0 checklist? I may be able to pitch for Gleam or simple enough Erlang tasks

**Comment by ollien** [comment]
TL;DR for onlookers: this bug stems from trying to load a table after already creating it, which is invalid even in plain `ets`. The workaround for this issue is to 

1. Create any ETS table with a different name
2. Call `uset.file2tab`

---

Hi @Michael-Mark-Edu 

Should you ever come back to this project, I've done some analysis that should help you out.

The program in the OP fails for a simple reason: it's trying to call `file2tab` on a table that already exists in memory. We can replicate this if we produce similar erlang

```
$ erl
Erlang/OTP 27 [erts-15.1.2] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit:ns]

Eshell V15.1.2 (press Ctrl+G to abort, type help(). for help)
1> Tab = ets:new(hello, [named_table]).
hello
2> ets:tab2file(Tab, "/tmp/hello").
ok
3> ets:file2tab("/tmp/hello").
{error,cannot_create_table}
```

This would imply that we could simply solve this by creating a Gleam program which just calls `uset.file2tab` (and not `uset.new`), but this fails in a slightly different way

```gleam
import gleam/io
import gleam/dynamic
import bravo/uset

pub fn main() {
  let path = "blah"

  let _ =
    uset.file2tab(path, False, dynamic.tuple2(dynamic.string, dynamic.string))
}
```

```
runtime error: Erlang error

An error occurred outside of Gleam.

erlang:error(Badarg)

stacktrace:
  ets.insert unknown source
  bravo_ffi.try_file2tab /home/nick/Documents/code/bravo/example/build/dev/erlang/bravo/_gleam_artefacts/bravo_ffi.erl:99
  bravo/uset.file2tab /home/nick/Documents/code/bravo/src/bravo/uset.gleam:158
```

It seems that `bravo_ffi.erl` is failing to insert to the `$BRAVOMETA` table, which is only created when `uset.new` is called. A workaround here would be to create a dummy table with `uset.new()`, then call `file2tab` to load your desired table.

Sidenote: the original issue fails on a call to `ets:first`, because the pattern match implicitly assumes that `ets:file2tab` will return `{ok, TableName}`, when in reality it can fail.

I think you can, at minimum, stop this from crashing with a patch like this

```diff
diff --git a/src/bravo_ffi.erl b/src/bravo_ffi.erl
index 3e9e7bb..e84254c 100644
--- a/src/bravo_ffi.erl
+++ b/src/bravo_ffi.erl
@@ -90,6 +90,10 @@ try_file2tab(Filename, Verify) ->
   case catch ets:file2tab(Filename, [{verify, Verify}]) of
     {'EXIT', {Reason, _}} ->
       {error, {erlang_error, atom_to_binary(Reason, utf8)}};
+
+    {error, Reason} ->
+      {error, {erlang_error, atom_to_binary(Reason, utf8)}};
+
     Other ->
       Name = element(2, Other),
       ets:insert('$BRAVOMETA', {Name, case is_tuple(ets:lookup(Name, ets:first(Name))) of
```

That said, the code in 5.0 seems to [already do this](https://github.com/Michael-Mark-Edu/bravo/blob/9ec7565fea973dac1192c266cd21c17cfeac256f/src/bravo_ffi.erl#L128) :)

**Comment by Michael-Mark-Edu** [comment]
I want to move away from using the `ErlangError(String)` subtype. The current build in main devolves to an `ErlangError` in this case.

I think in this situation, making `file2tab` return `Error(bravo.TableAlreadyExists)` should be sufficient resolution.

**Comment by Michael-Mark-Edu** [comment]
@ollien Thanks your your analysis! I don't have Gleam/Erlang debugging tools set up right now (if they even exist) so your deep dive into the bug was extremely useful!

**Comment by Michael-Mark-Edu** [comment]
Fixed by 4debb45
